### PR TITLE
fix: Websocket lifetime improvements

### PR DESCRIPTION
**Pull request by SandPod**
Collection of improvements related to WebSockets and their lifetime management.

### Changes:
- Correctly set session logs to current endpoint when connection is being closed.
- Only log failed websocket upgrade requests if logging verbose is enabled.
- Break out websocket request handling to separate file.
- Make server shutdown async and await all HTTP close calls.
- Close database pool on server shutdown.
- Close all websocket connections on server shutdown.

After these changes, the server can now be gracefully shut down. This was a required change in order to test the websocket lifetime management on the server. This also means we should no longer have to have the `exit(0)` call inside of the shutdown method.

### Guidance for review.
Best reviewed commit by commit to be able to follow the refactoring and break out of the WebSocket request handler.

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

None:

### CODE REVIEW

File: packages/serverpod/lib/src/server/server.dart [hunk 1]
```diff
@@ -546,6 +546,7 @@ class Server {
     Endpoint endpoint,
   ) async {
     try {
+      session.sessionLogs.currentEndpoint = endpoint.name;
```
Comment (Isakdl) [inline packages/serverpod/lib/src/server/server.dart]: Question: why is this necessary when closing the stream?

File: packages/serverpod/lib/src/relic/web_server.dart [hunk 2]
```diff
@@ -184,9 +184,9 @@ class WebServer {
   }
 
   /// Stops the webserver.
-  void stop() {
+  Future<void> stop() async {
     if (_httpServer != null) {
-      _httpServer!.close();
+      await _httpServer!.close();
```
Comment (Isakdl) [inline packages/serverpod/lib/src/relic/web_server.dart:189]: Suggestion: Change this to get a local variable so that we can remove the `!` cast.

File: packages/serverpod/lib/src/database/database_pool_manager.dart [hunk 3]
```diff
@@ -30,15 +41,16 @@ class DatabasePoolManager {
   DatabasePoolManager(
     SerializationManagerServer serializationManager,
     this.config,
-  ) {
+  ) : _poolSettings = pg.PoolSettings(
+          maxConnectionCount: 10,
```
Comment (Isakdl) [inline packages/serverpod/lib/src/database/database_pool_manager.dart:45]: Question: This is not relevant for this PR but is this a configuration we should expose for our users? It might be nice to be able to adjust this.

File: packages/serverpod/lib/src/server/server.dart [hunk 4]
```diff
@@ -237,8 +243,16 @@ class Server {
         return;
       }
       webSocket.pingInterval = const Duration(seconds: 30);
```
Comment (Isakdl) [inline packages/serverpod/lib/src/server/server.dart:245]: Question: Should we expose this setting for the end user? If we do we give a little more power for people to either reduce network traffic or increase confidence in that clients are alive, might be relevant depending on what you are building.

File: tests/serverpod_test_server/test_integration/websocket_connection_test.dart [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    tearDown(() async {
+      await server.shutdown(exitProcess: false);
+      await webSocket.sink.close();
+    });
+
+    test('when server is stopped then socket is closed.', () async {
+      webSocket.stream.listen((event) {
+        // Listen to the to keep it open.
```
Comment (Isakdl) [inline tests/serverpod_test_server/test_integration/websocket_connection_test.dart:26]: Nit: Listen to the **??** to keep it open.

**Review by Isakdl — APPROVED** [review]
LGTM with small changes.

I have a few questions, but you take the decision on if they are relevant to deal with.

File: packages/serverpod/lib/src/server/server.dart (hunk 1 above)
Comment (SandPod) [reply packages/serverpod/lib/src/server/server.dart]: The current endpoint is what is used for logging, so if some logging happens when we close the stream these logs will be reported for the wrong endpoint.

File: packages/serverpod/lib/src/database/database_pool_manager.dart (hunk 3 above)
Comment (SandPod) [reply packages/serverpod/lib/src/database/database_pool_manager.dart:45]: I agree with this! :)

File: packages/serverpod/lib/src/server/server.dart (hunk 4 above)
Comment (SandPod) [reply packages/serverpod/lib/src/server/server.dart:245]: I agree with this as well! Would be nice to have handles for these things :)
